### PR TITLE
[SPARK-21534][SQL][PySpark] PickleException when creating dataframe from python row with empty bytearray

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

`PickleException` is thrown when creating dataframe from python row with empty bytearray

    spark.createDataFrame(spark.sql("select unhex('') as xx").rdd.map(lambda x: {"abc": x.xx})).show()

    net.razorvine.pickle.PickleException: invalid pickle data for bytearray; expected 1 or 2 args, got 0
    	at net.razorvine.pickle.objects.ByteArrayConstructor.construct(ByteArrayConstructor.java
        ...

`ByteArrayConstructor` doesn't deal with empty byte array pickled by Python3.

## How was this patch tested?

Added test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @HyukjinKwon

**Comment by HyukjinKwon** [comment]
For me, it takes a while to double check this. Will try to help double check within this week.

**Comment by viirya** [comment]
Thanks @HyukjinKwon

File: core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala [hunk 1]
```diff
@@ -35,6 +35,16 @@ import org.apache.spark.rdd.RDD
 
 /** Utilities for serialization / deserialization between Python and Java, using Pickle. */
 private[spark] object SerDeUtil extends Logging {
+  class ByteArrayConstructor extends net.razorvine.pickle.objects.ByteArrayConstructor {
+    override def construct(args: Array[Object]): Object = {
+      // Deal with an empty byte array pickled by Python 3.
+      if (args.length == 0) {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala]: I see. It looks quite straightforward.  I checked in Python 3:

```
>>> import pickle
>>> import pickletools
>>> print(pickletools.dis(pickle.dumps(bytearray())))
    0: \x80 PROTO      3
    2: c    GLOBAL     'builtins bytearray'
   22: q    BINPUT     0
   24: )    EMPTY_TUPLE
   25: R    REDUCE
   26: q    BINPUT     1
   28: .    STOP
```

which, up to my knowledge, gives new object[0] for `args`.

**Comment by HyukjinKwon** [comment]
LGTM

@ueshin, could you double check this one when you have some time?

File: core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala (hunk 1 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala]: I also checked `pickle.dumps(..., protocol=0 - 4)` just in case.

**Review by ueshin — COMMENTED** [review]
LGTM except for some minor comments.

File: python/pyspark/sql/tests.py [hunk 2]
```diff
@@ -2480,6 +2480,11 @@ def assertCollectSuccess(typecode, value):
                 a = array.array(t)
                 self.spark.createDataFrame([Row(myarray=a)]).collect()
 
+    # test for SPARK-21534
+    def test_empty_bytearray(self):
+        rdd = self.spark.sql("select unhex('') as xx").rdd.map(lambda x: {"abc": x.xx})
+        self.spark.createDataFrame(rdd).collect()
```
[the remark this answers is not in the document]
Comment (viirya) [reply python/pyspark/sql/tests.py]: Yes. Added.

File: core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala [hunk 3]
```diff
@@ -35,6 +35,16 @@ import org.apache.spark.rdd.RDD
...
+        Array.empty[Byte]
```
[the remark this answers is not in the document]
Comment (viirya) [reply core/src/main/scala/org/apache/spark/api/python/SerDeUtil.scala]: Ok.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**